kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/ColumnMapping.java

```diff
@@ -343,18 +343,19 @@ private static Optional<Metadata> assignColumnIdAndPhysicalName(
                         .getConfiguration()
                         .getOrDefault(COLUMN_MAPPING_MAX_COLUMN_ID_KEY, "0")),
                 findMaxColumnId(oldSchema)));
-
-    StructType newSchema = new StructType();
-    for (StructField field : oldSchema.fields()) {
-      newSchema =
-          newSchema.add(
-              transformAndAssignColumnIdAndPhysicalName(
-                  assignColumnIdAndPhysicalNameToField(
-                      field, maxColumnId, isNewTable, useColumnIdForPhysicalName),
-                  maxColumnId,
-                  isNewTable,
-                  useColumnIdForPhysicalName));
-    }
+    SchemaIterable schemaIterable = new SchemaIterable(oldSchema);
+    schemaIterable
+        .mutableStream()
+        .filter(SchemaIterable.MutableSchemaElement::isStructField)
+        .forEach(
+            mutableElement ->
+                mutableElement.updateField(
```

Comment:
We could avoid some object churn here by modifying assignColumnIdAndPhysicalNameNameToField to only return a value if there is in fact a change.

Reply:
done.